spec/weak-ref.html

```diff
@@ -136,19 +136,24 @@ <h1>Properties of WeakRef Instances</h1>
       internal slot.
     </p>
   </emu-clause>
+</emu-clause>
 
-  <emu-clause id="sec-collection-modifications">
-    <h1>Modifications to collection type definitions</h1>
-    <p>WeakMap and WeakSet keys are not kept alive just because a WeakRef points to them.</p>
-
-    <emu-clause id="sec-weakmap">
-      <h1>WeakMap Objects</h1>
-      <p>WeakMap objects are collections of key/value pairs where the keys are objects and values may be arbitrary ECMAScript language values. A WeakMap may be queried to see if it contains a key/value pair with a specific key, but no mechanism is provided for enumerating the objects it holds as keys. If an object that is being used as the key of a WeakMap key/value pair is only reachable by following a chain of references that start within that WeakMap, <ins>or if all references to a ECMAScript object are from a [[Target]] field or internal slot,</ins> then that key/value pair is inaccessible and is automatically removed from the WeakMap. WeakMap implementations must detect and remove such key/value pairs and any associated resources.</p>
-    </emu-clause>
+<emu-clause id="sec-collection-modifications">
+  <h1>Modifications to collection type definitions</h1>
+  <emu-note type=editor>
+    WeakMap and WeakSet keys are not kept alive just because a WeakRef points
+    to them. This proposal rephrases the definition of WeakMaps and WeakSets
+    to explain their observable effects on garbage collection, rather than
+    specifying operationally that non-live keys or members are deleted.
+  </emu-note>
+
+  <emu-clause id="sec-weakmap">
+    <h1>WeakMap Objects</h1>
+    <p>WeakMap objects are collections of key/value pairs where the keys are objects and values may be arbitrary ECMAScript language values. A WeakMap may be queried to see if it contains a key/value pair with a specific key, but no mechanism is provided for enumerating the objects it holds as keys. <ins>There is no way for ECMAScript programs to observe the presence of a key in a WeakMap without a reference to that key. This implies that, if an object is not live, and it is present as a WeakMap key, then the implementation may unobservably remove the key-value pair from the WeakMap.</ins> <del>If an object that is being used as the key of a WeakMap key/value pair is only reachable by following a chain of references that start within that WeakMap, then that key/value pair is inaccessible and is automatically removed from the WeakMap. WeakMap implementations must detect and remove such key/value pairs and any associated resources.</del></p>
```

Comment:
This does seem to relax the existing requirement that unreachable keys and associated value **must** be removed:
> ~WeakMap implementations must detect and remove such key/value pairs and any associated resources.~

> the implementation may unobservably remove the key-value pair from the WeakMap

Of course the removal was unobservable before Weakrefs, but a non-collecting engine may now be compliant when it wasn't before.

Reply:
Yes, this relaxes the requirement's wording, but the effect is observably equivalent. I don't think the previous wording was really optimal, and I am proposing this change, as long as we are cleaning up the definitions around liveness.